hal/stm32f103.c

```diff
@@ -0,0 +1,324 @@
+#define RCC_CFGR_HPRE_DIV_NONE      (0 << 4)
+#define RCC_CFGR_HPRE_DIV_2         (8 << 4)
+#define RCC_CFGR_HPRE_DIV_4         (9 << 4)
+
+#define RCC_CFGR_PPRE1_MASK         0x0700
+#define RCC_CFGR_PPRE1_DIV_NONE     (0 << 8)
+#define RCC_CFGR_PPRE1_DIV_2        (4 << 8)
+#define RCC_CFGR_PPRE1_DIV_4        (5 << 8)
+
+#define RCC_CFGR_PPRE2_MASK         0x3800
+#define RCC_CFGR_PPRE2_DIV_NONE     (0 << 11)
+#define RCC_CFGR_PPRE2_DIV_2        (4 << 11)
+#define RCC_CFGR_PPRE2_DIV_4        (5 << 11)
+
+#define PLL_FULL_MASK (0x003F0000)
+#define RCC_CFGR_PLLSRC             (1 << 22)
+#define RCC_CFGR_PLLMUL_MUL_9       (7 << 18)
+
+/*** FLASH ***/
+#define RCC_APB1ENR         (*(volatile uint32_t *)(RCC_BASE + 0x1C))
+#define RCC_APB1ENR_PWREN   (1 << 28)
+
+#define FLASH_BASE          (0x40022000U)
+#define FLASH_ACR           (*(volatile uint32_t *)(FLASH_BASE + 0x00U))
+#define FLASH_KEYR          (*(volatile uint32_t *)(FLASH_BASE + 0x04U))
+#define FLASH_SR            (*(volatile uint32_t *)(FLASH_BASE + 0x0CU))
+#define FLASH_CR            (*(volatile uint32_t *)(FLASH_BASE + 0x10U))
+#define FLASH_AR            (*(volatile uint32_t *)(FLASH_BASE + 0x14U))
+
+#define FLASH_MAX_SZ        (128*1024) /* only low- and medium-density devices supported */
+#define FLASH_PAGE_SZ       (1024)
+
+/* Register values */
+#define FLASH_ACR_ENABLE_PRFT                 (1 << 4)
+
+#define FLASH_SR_BSY                          (1 << 0)
+
+#define FLASH_CR_LOCK                         (1 << 7)
+#define FLASH_CR_STRT                         (1 << 6)
+#define FLASH_CR_PER                          (1 << 1)
+#define FLASH_CR_PG                           (1 << 0)
+
+#define FLASH_KEY1                            (0x45670123U)
+#define FLASH_KEY2                            (0xCDEF89ABU)
+
+static void RAMFUNCTION flash_set_waitstates(int waitstates)
+{
+    FLASH_ACR |=  waitstates | FLASH_ACR_ENABLE_PRFT;
+}
+
+int RAMFUNCTION valid_flash_area(uint32_t address, int len)
+{
+    if (len <= 0 || len > FLASH_MAX_SZ)
+        return -1;
+    if (address < ARCH_FLASH_OFFSET || address >= (ARCH_FLASH_OFFSET + FLASH_MAX_SZ))
+        return -1;
+    if ((address + len) > (ARCH_FLASH_OFFSET + FLASH_MAX_SZ))
+        return -1;
+
+    return 0;
+}
+
+static RAMFUNCTION inline void flash_wait_complete(void)
+{
+    while ((FLASH_SR & FLASH_SR_BSY) == FLASH_SR_BSY)
+        ;
+}
+
+static int RAMFUNCTION flash_erase_page(uint32_t address)
+{
+    const uint32_t end = address + FLASH_PAGE_SZ;
+
+    flash_wait_complete();
+
+    FLASH_CR |= FLASH_CR_PER;
+    FLASH_AR = address;
+    FLASH_CR |= FLASH_CR_STRT;
+    flash_wait_complete();
+    FLASH_CR &= ~FLASH_CR_PER;
+
+    /* verify erase */
+    while (address < end) {
+        if (*(uint32_t*)address != (uint32_t)~0)
+            return -1;
+
+        address += 4;
+    }
+
+    return 0;
+}
+
+int RAMFUNCTION hal_flash_write(uint32_t address, const uint8_t *data, int len)
```

Comment:
Don't worry at all about erasing the page before write. This task is performed by wolfBoot, that will always call erase before writing. If the flash memory does not support consecutive writes after erase, just add NVM_FLASH_WRITEONCE=1 to the example config, like for similar STM32 targets.

Please remove any erase commands issued in hal_flash_write.

Reply:
Ah, I see! I couldn't figure out if this was the case so I opted for the safe approach. This makes the implementation much simpler!

Reply:
So, the STM32F1 do support consecutive writes after erase, but it only supports 16 bit writes. I'm trying out an idea to see if a read-modify-write approach to unaligned writes would work but will take some more testing to verify. I currently have a single LED as my only debugging tool for Wolfboot, but I might be able to do some tests with Zephyr or arduino to test it out. It will take a little more time, unfortunately

Reply:
I've done some testing using Zephyr, and I can confirm that writing to flash only works when:

- the write is a 2 byte operation
- the area is 0xFFFF (erased)

Writing 2 bytes to an area that is _not_ erased does nothing, even if it's `0xFFaa` or `0xbbFF` and you write `0xccaa` or `0xccFF`.

I realize that I misunderstood what `NVM_FLASH_WRITEONCE=1` meant. I thought it meant that writes to a whole _sector_ had to be written at once. After reading the docs better I realize that it's meant to address the exact issue I just described. I'll clean up the code and add `NVM_FLASH_WRITEONCE=1` to the example config.

Reply:
The code should be good for review now.

I've received the Contributor Agreement form, but haven't got around to reading or signing it yet.

Reply:
The form should be signed now